### PR TITLE
✨ Computational backend: connect to resource tracking via RabbitMQ (🗃️, ⚠️)

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
This PR let now the director-v2 send messages when computational services are started/stopped and while they are running (heartbeat concept)

- [x] passes correct simcore_user_agent to instrumentation RabbitMQ messages when running computational services
- [x] publishes Resource Usage Tracker messages events when start/stop computational services
- [x] publishes Resource Usage Tracker heartbeat events when running computational services
- [x] (⚠️Devops) new ENV ```SERVICE_TRACKING_HEARTBEAT``` in date time [format](https://docs.pydantic.dev/latest/usage/types/datetime/), defines the interval with which a heartbeat is sent into RabbitMQ while a computational service is running **defaults to 60 seconds**, this ENV is global as it will be used in other services (such as RUTH)
- [x] 🗃️*comp_tasks* table enhanced with ```last_heartbeat```, ```created```, ```modified``` timestamps
- [x] added tests to check for correct publication of the RabbitMQ messages
- [x] 🗃️ new RabbitMQ exchange to send messages about started,stopped and heartbeat messages is used from the computational scheduler

various:
- [x] removed RunningState.RETRY state as unused and confusing
- [x] fixed running postgres-database with external database

## Related issue/s
- solves https://github.com/ITISFoundation/osparc-simcore/issues/4523

### CODE REVIEW

**Review by mrnicegyu11 — COMMENTED** [review]
tldr

green light for the ops changes thanks a lot for the nice summary in the PR text :--)

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/afc752d10a6c_add_waiting_for_resources_enum_field.py [hunk 1]
```diff
@@ -22,13 +22,15 @@ def upgrade():
 
     conn = op.get_bind()
     result = conn.execute(
-        f"SELECT * FROM pg_enum WHERE enumtypid = (SELECT oid FROM pg_type WHERE typname = '{enum_type_name}') AND enumlabel = '{new_value}'"
+        sa.DDL(
+            f"SELECT * FROM pg_enum WHERE enumtypid = (SELECT oid FROM pg_type WHERE typname = '{enum_type_name}') AND enumlabel = '{new_value}'"
+        )
     )
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/afc752d10a6c_add_waiting_for_resources_enum_field.py:28]: no. that is an implicit conversion. I am just adding the ```sa.DDL``` to make sqlalchemy 2.0 happy

**Review by matusdrobuliak66 — APPROVED** [review]
Looking good! :+1:

File: services/director-v2/src/simcore_service_director_v2/core/settings.py [hunk 2]
```diff
@@ -77,6 +78,8 @@ class PlacementConstraintStr(ConstrainedStr):
 
 
 class VFSCacheMode(str, Enum):
+    __slots__ = ()
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/core/settings.py:81]: this is a Ruff recommendation, see https://beta.ruff.rs/docs/rules/no-slots-in-str-subclass/

**Review by YuryHrytsuk — APPROVED** [review]
*Thinking aloud*

If we want to be fully transparent with env and avoid implicit defaults (in code), then we want to have no defaults in code and provide values transparently via `.env` files :thinking: 

*I am still not sure if we want to go in this direction

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py [hunk 3]
```diff
@@ -246,88 +327,146 @@ async def _get_changed_tasks_from_backend(
...
-        user_id: UserID,
-        project_id: ProjectID,
-        changed_tasks: list[tuple[_Previous, _Current]],
-    ) -> None:
-        for previous, current in changed_tasks:
-            if current.state is RunningState.STARTED or (
-                previous.state in WAITING_FOR_START_STATES
-                and current.state in COMPLETED_STATES
-            ):
-                message = InstrumentationRabbitMessage.construct(
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py:282]: Could you please also replace the following calls in the following as well, since you are changing this:

https://github.com/ITISFoundation/osparc-simcore/blob/54970e3cb376515baa8e174db77d6c56be339b27/services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py#L112

https://github.com/ITISFoundation/osparc-simcore/blob/54970e3cb376515baa8e174db77d6c56be339b27/services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events_utils.py#L351

File: services/director-v2/src/simcore_service_director_v2/utils/rabbitmq.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        service_type=task.node_class.value,
+        service_key=task.image.name,
+        service_tag=task.image.tag,
+        result=task_final_state,
+        simcore_user_agent=simcore_user_agent,
+    )
+    await rabbitmq_client.publish(message.channel_name, message)
+
+
+async def publish_service_resource_tracking_started(  # noqa: PLR0913
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/utils/rabbitmq.py:66]: it would be great if these were part of some library so that I could reuse them in the dy-sidecar.

**Review by GitHK — APPROVED** [review]
Please consider my comments regarding the moving of some code.

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py (hunk 3 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py:282]: not sure I follow what you want here?

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py [hunk 5]
```diff
@@ -246,88 +327,146 @@ async def _get_changed_tasks_from_backend(
...
-                    service_uuid=current.node_id,
-                    service_type=NodeClass.COMPUTATIONAL.value,
-                    service_key=current.image.name,
-                    service_tag=current.image.tag,
-                    simcore_user_agent=UNDEFINED_DEFAULT_SIMCORE_USER_AGENT_VALUE,
+    async def _process_reverted_tasks(self, tasks: list[CompTaskAtDB]) -> None:
+        _logger.warning(
+            "Tasks were reverted, this should not happen! TIP: Check %s",
+            f"{[t.job_id for t in tasks]}",
+        )
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py:416]: if it becomes an issue yes. anyway this part of the code will still undergo some refactoring soon.

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py [hunk 6]
```diff
@@ -39,34 +39,71 @@
...
+    publish_service_resource_tracking_started,
+    publish_service_started_metrics,
 )
 from ..db.repositories.comp_pipelines import CompPipelinesRepository
 from ..db.repositories.comp_runs import CompRunsRepository
 from ..db.repositories.comp_tasks import CompTasksRepository
 
 _logger = logging.getLogger(__name__)
 
+_UNDEFINED_METADATA = "undefined"
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py:67]: I moved this to the constants inside the director-v2.
This value has nothing to do with what you do in the dynamic services AFAIK. these have to do with computational run metadata that also contain other stuff.
Therefore if you add something similar, you are free to re-use it of course, but I do not see the point in putting this in service-lib for instance.

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py (hunk 3 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py:282]: you mean using the construct instead of parsing what should not be parsed?
